src/handlers/records-query.ts

```diff
@@ -26,39 +27,38 @@ export class RecordsQueryHandler implements MethodHandler {
       return messageReplyFromError(e, 400);
     }
 
+    let recordsWrites: RecordsWriteMessageWithOptionalEncodedData[];
+
     // if this is an anonymous query, query only published records
     if (recordsQuery.author === undefined) {
-      let recordsWrites = await this.fetchPublishedRecords(tenant, recordsQuery);
-      recordsWrites = await sortRecords(recordsWrites, recordsQuery.message.descriptor.dateSort);
-
-      const entries = RecordsQueryHandler.removeAuthorization(recordsWrites);
-      return {
-        status: { code: 200, detail: 'OK' },
-        entries
-      };
-    }
+      recordsWrites = await this.fetchPublishedRecords(tenant, recordsQuery);
+    } else {
+      // authentication
+      try {
+        await authenticate(message.authorization!, this.didResolver);
+      } catch (e) {
+        return messageReplyFromError(e, 401);
+      }
 
-    // authentication
-    try {
-      await authenticate(message.authorization!, this.didResolver);
-    } catch (e) {
-      return messageReplyFromError(e, 401);
+      if (recordsQuery.author === tenant) {
+        recordsWrites = await this.fetchRecordsAsOwner(tenant, recordsQuery);
+      } else {
+        recordsWrites = await this.fetchRecordsAsNonOwner(tenant, recordsQuery);
+      }
     }
 
-    let recordsWrites: RecordsWriteMessageWithOptionalEncodedData[];
-    if (recordsQuery.author === tenant) {
-      recordsWrites = await this.fetchRecordsAsOwner(tenant, recordsQuery);
-    } else {
-      recordsWrites = await this.fetchRecordsAsNonOwner(tenant, recordsQuery);
+    let paginationMessageCid: string|undefined;
+    const lastRecord = recordsWrites.at(-1);
+    if (lastRecord) {
```

Comment:
Should we only return cursor if paging is specified in the query? Going even further would be to only return cursor if we KNOW that there are more entries to be returned, but this would further suggest that the cursor be returned from the query call: e.g. 

```ts
  query(
    tenant: string,
    filter: Filter,
    dateSort?: MessageSort,
    pagination?: Pagination,
    options?: MessageStoreOptions
  ): Promise<{ messages: GenericMessage[], paginationMessageCid: string }>;
  ```

Reply:
Hrm... I think you're right we should only return the cursor if pagination is specified.

But I'm not sure about going the next step as only including it if we know there are more records. I think of it as just a cursor, not really knowing knowledge of what's after or before it.

Either way, I do think that having `paginationMessageCid` returned with `query` makes the most sense.
We would also at some point want to expose pagination on the other Query interfaces and it feels right for it to come from the query method of the store.

Reply:
@LiranCohen, responding to the money wrench you brought up regarding our code using multiple queries internally [(code)](https://github.com/TBD54566975/dwn-sdk-js/blob/de43a09c70eaef0aa5dc7520c5fdaae631dbff63/src/handlers/records-query.ts#L111) to serve the external query, complicating pagination logic...

Super glad that you catch this, we need to have a test that reproduce this so that we can confirm the scenario is handled and more importantly we don't regress.

The way I see it, there are two ways to approach this:

1. Combine multiple query into one somehow then pass it to the message store to execute. This requires message store to support OR predicates, something we've deliberately avoided to keep things simple and I still think it is the right decision given the resourcing we have. This is a can of worm I think we should avoid.
2. Add trimming logic accordingly in the query handler. Yes, we will need to incur 3 DB calls instead of 1 like we already do, but functionally it should/can work with some juggling, but avoidance of complexity increase is a worthwhile trade off IMO. This is what I am leaning towards.

Back to the discussion on whether we should have the message store query return the cursor, I am leaning towards it only if it's a low-enough hanging fruit, what I think is useful ultimately is a hint as to "if there are more message to fetch" or not, so technically a boolean would suffice. But keeping the code as is: externally always return a `messageCid` even though there might not be new messages is NOT the end of the world if you'd prefer this behavior, the ramification is "only" that the caller will need to make one extra call to know if there are more messages.

Reply:
Reached agreement to go for option 3: introduce query array and let the store do the pagination and sorting.